passes/source2il.nim

```diff
@@ -0,0 +1,87 @@
+## This implements a temporary pass for translating the parse-tree of the
+## source language into the currently highest-level intermediate language.
+## It's there so that development of the source language can already commence
+## while the the intermediate languages are still being developed.
+##
+## The focus is on correctness. Performance, code quality, and overall
+## architecture (of this module) are of secondary concern.
+
+import
+  passes/[builders, spec, trees],
+  vm/[utils]
+
+from passes/spec_source import nil
+
+type
+  SourceKind = spec_source.NodeKind
+  InTree     = PackedTree[SourceKind]
+  Node       = TreeNode[NodeKind]
+
+  TypeKind* = enum
+    tkInt
+    tkFloat
+
+using
+  bu: var Builder[NodeKind]
+
+proc exprToIL(t: InTree, n: NodeIndex, bu): TypeKind =
+  case t[n].kind
+  of SourceKind.IntVal:
+    bu.add Node(kind: IntVal, val: t[n].val)
+    result = tkInt
+  of SourceKind.FloatVal:
+    bu.add Node(kind: FloatVal, val: t[n].val)
+    result = tkFloat
+  of spec_source.AllNodes - spec_source.ExprNodes:
+    unreachable()
+
+proc exprToIL*(t: InTree): (TypeKind, PackedTree[NodeKind]) =
+  ## Translates the given source language expression to the highest-level IL.
+  ## Also returns the type of the expression.
+  var typ: TypeKind
+  block:
+    # we need to know the type of the expression, so we translate the
+    # expression using a temporary builder first
+    var bu = initBuilder[NodeKind]()
+    bu.subTree Module:
+      typ = exprToIL(t, NodeIndex(0), bu)
+    # XXX: this is bad. The Changeset/Builder API needs to support
+    #      placeholders, where one can insert placeholders, with the
+    #      actual subtree being created separately
```

Comment:
The ability to create placeholders and fill them in later is one thing that's sorely missing from NimSkull's `MirBuilder`. For lowering passes, it's *usually* not a problem, but especially `mirgen` or places where production and attribute inheritance are combined into one pass (such as is the case here) suffer from the lack of such API (the double-buffering scheme used by `MirBuilder` is an attempt to mitigate this, but it has its limits).

The interface I'm aiming for is the following:
```nim
var ph: Placeholder
bu.subTree Asgn:
  bu.add Node(...)
  ph = bu.placeholder()

bu.subTree ph, Call:
  ...
```

Reply:
The builder would need an enum that had one extra item to represent the placeholder, but otherwise seems pretty handy. It'd be nice if we didn't need the functionality, but after trying to rethink the analysis to avoid it, it seems impossible.